### DISCUSSION TITLE
Markdown render hooks

### DISCUSSION THREAD

**Discussion by stereobooster** [withastro/roadmap#690] [Proposal]
### Body

# Summary

Inspired by Hugo's [render-hooks](https://gohugo.io/templates/render-hooks/). Idea is to be able to provide custom Astro components to render some markdown elements, like:

- code
- header
- link
- image

# Background & Motivation

Basic on my experience working with Hugo this is a very nice feature. It allows to do neat tricks. Yes, most of this can be done with rehype plugins.

Similar proposals / issues:

- https://github.com/withastro/roadmap/discussions/423
- https://github.com/withastro/roadmap/discussions/488
- https://github.com/withastro/roadmap/discussions/346
- https://github.com/withastro/roadmap/discussions/267
- https://github.com/withastro/roadmap/discussions/424
- https://github.com/withastro/roadmap/discussions/329
- https://github.com/withastro/roadmap/discussions/491
- https://github.com/withastro/roadmap/discussions/180

# Goals

- by allowing customizing header
  - people can add [anchor links](https://github.com/valeriangalliat/markdown-it-anchor/issues/82) or change how slugs are generated
- by allowing customizing image
  - people can add lazy loading, srcset (which already supported by Astro), [lqip](https://github.com/transitive-bullshit/lqip-modern)
- by allowing customizing link
  - people can add `rel="nofollow" target="_blank"` to external links
- by allowing customizing code
  - people can use different syntax highlighter
  - pass different configuration to syntax highlighter
  - implement [diagram rendering](https://code.juliancataldo.com/component/astro-diagram/)

There are already enough plugins with custom components it would just metter of plugin things together.

# Example

```astro
---
// src/components/Link.astro

import { type LinkProps } from 'astro:markdown';
type Props = LinkProps;
---

<a href={Astro.props.href} target="_blank" rel="nofollow">
  {Astro.children}
</a>
```

Related: https://github.com/johnzanussi/astro-mdx-code-blocks#readme

**Comment by delucis** [comment]
You might want to check out MDX, which supports this pattern: https://docs.astro.build/en/guides/markdown-content/#custom-components-with-imported-mdx

**Reply by stereobooster** [reply]
Thanks. I was aware that MDX can do that, but somehow I didn't realize that I can apply it in Astro. Maybe because most documentation shows examples of JSX tags and never mentions this option.
